model/src/applications/command/option.rs

```diff
@@ -0,0 +1,296 @@
+                envelope.choices.take().unwrap_or_default(),
+                envelope.into(),
+            ))),
+            CommandOptionType::Boolean => CommandOption::Boolean(envelope.into()),
+            CommandOptionType::User => CommandOption::User(envelope.into()),
+            CommandOptionType::Channel => CommandOption::Channel(envelope.into()),
+            CommandOptionType::Role => CommandOption::Role(envelope.into()),
+            CommandOptionType::Mentionable => CommandOption::Mentionable(envelope.into()),
+        }
+    }
+}
+
+/// Data supplied to a [`CommandOption`] of type [`Boolean`], [`User`],
+/// [`Channel`], or [`Role`].
+///
+/// [`Boolean`]: CommandOption::Boolean
+/// [`User`]: CommandOption::User
+/// [`Channel`]: CommandOption::Channel
+/// [`Role`]: CommandOption::Role
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct BaseCommandOptionData {
+    /// Description of the option. It must be 100 characters or less.
+    pub description: String,
+    /// Name of the option. It must be 32 characters or less.
+    pub name: String,
+    /// Whether the option is required to be completed by a user.
+    #[serde(default)]
+    pub required: bool,
+}
+
+impl From<CommandOptionEnvelope> for BaseCommandOptionData {
+    fn from(envelope: CommandOptionEnvelope) -> Self {
+        Self {
+            description: envelope.description,
+            name: envelope.name,
+            required: envelope.required,
+        }
+    }
+}
+
+/// Data supplied to a [`CommandOption`] of type [`SubCommand`] or
+/// [`SubCommandGroup`].
+///
+/// [`SubCommand`]: CommandOption::SubCommand
+/// [`SubCommandGroup`]: CommandOption::SubCommandGroup
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct OptionsCommandOptionData {
+    /// Description of the option. It must be 100 characters or less.
+    pub description: String,
+    /// Name of the option. It must be 32 characters or less.
+    pub name: String,
+    /// Used for specifying the nested options in a [`SubCommand`] or
+    /// [`SubCommandGroup`].
+    ///
+    /// [`SubCommand`]: CommandOptionType::SubCommand
+    /// [`SubCommandGroup`]: CommandOptionType::SubCommandGroup
+    #[serde(default)]
+    pub options: Vec<CommandOption>,
+    /// Whether the option is required to be completed by a user.
+    #[serde(default)]
+    pub required: bool,
+}
+
+impl From<(Vec<CommandOption>, BaseCommandOptionData)> for OptionsCommandOptionData {
+    fn from(opt: (Vec<CommandOption>, BaseCommandOptionData)) -> Self {
+        Self {
+            description: opt.1.description,
+            name: opt.1.name,
+            options: opt.0,
+            required: opt.1.required,
+        }
+    }
+}
+
+impl OptionsCommandOptionData {
+    fn base(&self) -> BaseCommandOptionData {
+        BaseCommandOptionData {
+            description: self.description.clone(),
+            name: self.name.clone(),
```

Comment:
We should take ownership of self here to avoid a required clone. If the user wants to keep the instance of `OptionsCommandOptionData` while getting a `BaseCommandOptionData` then they can clone `OptionsCommandOptionData`.

Reply:
These aren't public methods, this is just for serialization. I think these impls can be removed if custom ser is done in a different way